fms_extras/models/speculator.py

```diff
@@ -0,0 +1,124 @@
+import math
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from fms.modules.layernorm import LayerNormParameterized
+
+class MLP_Speculator(nn.Module):
+    """
+    This is a simple MLP-based speculator that functions similarly to Medusa, ingesting context via
+    the final embedding vector from the base model. However, this model also conditions on previously
+    predicted tokens, similarly to an RNN, allowing it to generate better-quality n-grams. 
+
+    The architecture is as flat and simple as possible: for each prediction head, the current 
+    state vector is projected into a new latent space and added to the previous token's embedding. 
+    This sum goes through layernorm and activation, forming the new state vector. This state predicts
+    the next token (or set of candidate tokens) for the current head, and then is passed on to the next.
```

Comment:
nit: in diffs on github, the text will be easier to read if each lines is shorter (short enough to not need to wrap in a side-by-side window).

long enough that whole paragraphs are wrapped works too though then they become harder to comment on